c++/nda/algorithms.hpp

```diff
@@ -199,6 +200,73 @@ namespace nda {
     return fold(std::multiplies<>{}, a, get_value_t<A>{1});
   }
 
+  /**
+   * @brief Hadamard product of two nda::Array objects.
+   * 
+   * @tparam A nda::Array type.
+   * @tparam B nda::Array type.
+   * @param a nda::Array object.
+   * @param b nda::Array object.
+   * @return nda::Array containing the elementwise product of the two input arrays.
+   */
+  template <Array A, Array B>
+    requires(nda::get_rank<A> == nda::get_rank<B>)
+  [[nodiscard]] constexpr auto hadamard(A &&a, B &&b) {
+    assert(a.shape() == b.shape());
+    return nda::map([](auto const &x, auto const &y) { return x * y; })(std::forward<A>(a), std::forward<B>(b));
+  }
+
+  /**
+   * @brief Hadamard product of two std::array objects.
+   * 
+   * @tparam T Data type of the arrays.
+   * @tparam R Size of the arrays.
+   * @param a std::array object.
+   * @param b std::array object.
+   * @return std::array containing the elementwise product of the two input arrays.
+   */
+  template <typename T, size_t R>
+  [[nodiscard]] constexpr auto hadamard(std::array<T, R> const &a, std::array<T, R> const &b) {
+    assert(a.size() == b.size());
+    return a * b;
+  }
+
+  /**
+   * @brief Hadamard product of two std::vector objects.
+   * 
+   * @tparam T Data type of the first input vector.
+   * @tparam U Data type of the second input vector.
+   * @param a std::vector object.
+   * @param b std::vector object.
+   * @return std::vector containing the elementwise product of the two input vectors.
+   */
+  template <typename T, typename U>
+  [[nodiscard]] constexpr auto hadamard(std::vector<T> const &a, std::vector<U> const &b) {
```

Comment:
I would expect a similar behavior between std::vector and std::array. Can you generalize the array case to allow `TU` type propagation in the same way? I.e. updating the `operator*`

Reply:
Checks out! Should be fixed by b3df66ac48ba3d045d2c5b9ed871e319388bbe05. What do you think?